feat: port pyclient build config to pyproject.toml

dshemetov opened the pull request:
### Summary:

I hoped the more modern config's [dynamic metadata](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata) would work, but it only works if `__version__` is defined explicitly in `__init__.py`, not imported from `delphi_epidata.py` (importing `delphi_epidata.py` errors at build-time because the dependencies in that file aren't installed in the build environment (just setuptools)).

So, this doesn't come with a useful feature, it's just a lateral move to the more modern packaging config. 

On the plus side, I learned a lot by reading parts of these guides:

- [PyPA Sample Project](https://github.com/pypa/sampleproject)
- [setuptools docs](https://setuptools.pypa.io/en/latest/index.html)
- [PyPA Packaging User Guide](https://packaging.python.org/en/latest/)

Might be handy in the future when we do #939 and other such changes.

### Prerequisites:

- [x] Unless it is a documentation hotfix it should be merged against the `dev` branch
- [x] Branch is up-to-date with the branch to be merged with, i.e. `dev`
- [x] Build is successful
- [x] Code is cleaned up and formatted

melange396 commented on src/client/packaging/pypi/pyproject.toml:
```diff
@@ -0,0 +1,118 @@
+# https://packaging.python.org/en/latest/tutorials/packaging-projects/#choosing-a-build-backend
+[build-system]
+# A list of packages that are needed to build your package:
+requires = ["setuptools>=61.0.0"] # REQUIRED if [build-system] table is used
```
what does that version get us?  the sample/template `pyproject.toml` [doesnt specify it](https://github.com/pypa/sampleproject/blob/a1f00056ebc65d17ae69c5aba324ef4f0382138a/pyproject.toml#L12)

dshemetov replied:
Ah yea, it's not necessary here, I'll remove it. It's needed to support the dynamic version feature (see the callout box in bullet 1 [here](https://packaging.python.org/en/latest/guides/single-sourcing-package-version/) "As of the release of setuptools 61.0.0...").

melange396 commented on src/client/packaging/pypi/pyproject.toml:
```diff
@@ -0,0 +1,118 @@
+[project]
+# This is the name of your project. The first time you publish this
+# package, this name will be registered for you. It will determine how
+# users can install this project, e.g.:
+#
+# $ pip install sampleproject
+#
+# And where it will live on PyPI: https://pypi.org/project/sampleproject/
+#
+# There are some restrictions on what makes a valid project name
+# specification here:
+# https://packaging.python.org/specifications/core-metadata/#name
+name = "delphi_epidata" # REQUIRED, is the only field that cannot be marked as dynamic.
+
+# Versions should comply with PEP 440:
+# https://www.python.org/dev/peps/pep-0440/
+#
+# For a discussion on single-sourcing the version, see
+# https://packaging.python.org/guides/single-sourcing-package-version/
+version = "4.1.23"
+
+# This is a one-line description or tagline of what your project does. This
+# corresponds to the "Summary" metadata field:
+# https://packaging.python.org/specifications/core-metadata/#summary
+description = "A programmatic interface to Delphi's Epidata API."
+
+# This is an optional longer description of your project that represents
+# the body of text which users will see when they visit PyPI.
+#
+# Often, this is the same as your README, so you can just read it in from
+# that file directly.
+#
+# This field corresponds to the "Description" metadata field:
+# https://packaging.python.org/specifications/core-metadata/#description-optional
+readme = "README.md"
+
+# Specify which Python versions you support. In contrast to the
+# 'Programming Language' classifiers in this file, 'pip install' will check this
+# and refuse to install the project if the version does not match. See
+# https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires
```
these are simple fields, all those comments arent necessary and make it harder to find relevant info
```suggestion
name = "delphi_epidata" # REQUIRED, is the only field that cannot be marked as dynamic.
version = "4.1.23"
description = "A programmatic interface to Delphi's Epidata API."
readme = "README.md"
```

melange396 commented on src/client/packaging/pypi/pyproject.toml:
```diff
@@ -0,0 +1,118 @@
+# 'Programming Language' classifiers in this file, 'pip install' will check this
+# and refuse to install the project if the version does not match. See
+# https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires
+requires-python = ">=3.8"
```
what is the earliest version we can use here?  im a little worried that this could be a breaking change for a lot of people, especially since we didnt specify a python version constraint before.

dshemetov replied:
So the client file claims that it works with Python 2 and the old setup.py file made no restrictions on Python version. It's probably best to remove this 👍

melange396 commented on src/client/packaging/pypi/pyproject.toml:
```diff
@@ -0,0 +1,118 @@
+# https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires
+requires-python = ">=3.8"
+
+# This is either text indicating the license for the distribution, or a file
+# that contains the license.
+# https://packaging.python.org/en/latest/specifications/core-metadata/#license
+license = { file = "LICENSE.txt" }
```
```suggestion
license = { file = "LICENSE" }
```
our license file doesnt have the txt suffix

melange396 commented on src/client/packaging/pypi/pyproject.toml:
```diff
@@ -0,0 +1,118 @@
+# https://packaging.python.org/en/latest/specifications/core-metadata/#license
+license = { file = "LICENSE.txt" }
+
+# This should be your name or the name of the organization who originally
+# authored the project, and a valid email address corresponding to the name
+# listed.
```
more unnecessary comments
```suggestion
```

melange396 commented on src/client/packaging/pypi/pyproject.toml:
```diff
@@ -0,0 +1,118 @@
+# listed.
+authors = [{ name = "David Farrow", email = "dfarrow0@gmail.com" }]
+
+# This should be your name or the names of the organization who currently
+# maintains the project, and a valid email address corresponding to the name
+# listed.
+# maintainers = [
+#     { name = "A. Great Maintainer", email = "maintainer@example.com" },
+# ]
```
```suggestion
maintainers = [
    { name = "Delphi Support", email = "delphi-support+pypi@andrew.cmu.edu" },
]
```

melange396 commented on src/client/packaging/pypi/pyproject.toml:
```diff
@@ -0,0 +1,118 @@
+#     { name = "A. Great Maintainer", email = "maintainer@example.com" },
+# ]
+
+# Classifiers help users find your project by categorizing it.
+#
+# For a list of valid classifiers, see https://pypi.org/classifiers/
+classifiers = [
+    # How mature is this project? Common values are
+    #   3 - Alpha
+    #   4 - Beta
+    #   5 - Production/Stable
+    # Pick your license as you wish
+    "License :: OSI Approved :: MIT License",
+
+    # Specify the Python versions you support here. In particular, ensure
+    # that you indicate you support Python 3. These classifiers are *not*
+    # checked by "pip install". See instead "requires-python" key in this file.
+    "Programming Language :: Python",
+    "Operating System :: OS Independent",
+    "Intended Audience :: Science/Research",
+    "Natural Language :: English",
+    "Topic :: Scientific/Engineering :: Bio-Informatics",
+]
```
get rid of more comments, add python v3 as one comment suggested.  i dont wanna claim this package is "stable" but im not sure if "beta" is appropriate either...  i wouldnt be upset if we just omitted the development status tag ¯\\\_(ツ)_/¯ 
```suggestion
# For a list of valid classifiers, see https://pypi.org/classifiers/
classifiers = [
    "Development Status :: 4 - Beta",
    "License :: OSI Approved :: MIT License",
    "Programming Language :: Python",
    "Programming Language :: Python :: 3",
    "Operating System :: OS Independent",
    "Intended Audience :: Science/Research",
    "Natural Language :: English",
    "Topic :: Scientific/Engineering :: Bio-Informatics",
]
```

melange396 commented on src/client/packaging/pypi/pyproject.toml:
```diff
@@ -0,0 +1,118 @@
+    "Topic :: Scientific/Engineering :: Bio-Informatics",
+]
+
+# This field lists other packages that your project depends on to run.
+# Any package you put here will be installed by pip when your project is
+# installed, so they must be valid existing projects.
+#
+# For an analysis of this field vs pip's requirements files see:
+# https://packaging.python.org/discussions/install-requires-vs-requirements/
+dependencies = ["aiohttp", "delphi-utils", "requests>=2.7.0", "tenacity"]
+
+
+# List URLs that are relevant to your project
+#
+# This field corresponds to the "Project-URL" and "Home-Page" metadata fields:
+# https://packaging.python.org/specifications/core-metadata/#project-url-multiple-use
+# https://packaging.python.org/specifications/core-metadata/#home-page-optional
+#
+# Examples listed include a pattern for specifying where the package tracks
+# issues, where the source is hosted, where to say thanks to the package
+# maintainers, and where to support the project financially. The key is
+# what's used to render the link text on PyPI.
```
more superfluous comments
```suggestion
dependencies = ["aiohttp", "delphi-utils", "requests>=2.7.0", "tenacity"]

```

melange396 reviewed (changes requested):
this looks great, i really like it!  there are waaaaaay too many comments in here tho....

melange396 reviewed (changes requested):
looks good, except i dont think this line is going to work any more: https://github.com/cmu-delphi/delphi-epidata/blob/69835d1d7795eaf9a710d9f4903fef22a07e8fdf/.github/workflows/release-helper.yml#L67

which probably also means updating this line:
https://github.com/cmu-delphi/delphi-epidata/blob/69835d1d7795eaf9a710d9f4903fef22a07e8fdf/.github/workflows/release-helper.yml#L60

and we should put an entry in https://github.com/cmu-delphi/delphi-epidata/blob/dev/src/client/packaging/pypi/CHANGELOG.md

dshemetov commented:
Good point, thanks! I swapped the commands over to the new ones ([sdist](https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#source-distributions) and [wheel](https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#pure-python-wheels)) and added the new `build` dependency to CI. 

Also updated the publish action based on hints from my IDE:

<img width="562" alt="image" src="https://github.com/cmu-delphi/delphi-epidata/assets/1810426/127dd428-4f8f-46c0-8580-b99b455e0e85">

And updated changelog too.

melange396 commented:
we can always fix the date on release day
